src/coreclr/interpreter/compiler.cpp

```diff
@@ -807,6 +856,94 @@ void InterpCompiler::EmitCode()
     m_compHnd->setBoundaries(m_methodInfo->ftn, m_ILToNativeMapSize, m_pILToNativeMap);
 }
 
+void InterpCompiler::BuildGCInfo(InterpMethod *pInterpMethod)
+{
+#ifdef FEATURE_INTERPRETER
+    InterpIAllocator* pAllocator = new (this) InterpIAllocator(this);
+    InterpreterGcInfoEncoder* gcInfoEncoder = new (this) InterpreterGcInfoEncoder(m_compHnd, m_methodInfo, pAllocator, Interp_NOMEM);
+    assert(gcInfoEncoder);
+
+    gcInfoEncoder->SetCodeLength(ConvertOffset(m_methodCodeSize));
+
+    uint32_t stackSlotCount = m_totalVarsStackSize / INTERP_STACK_SLOT_SIZE;
+    GcSlotId *slotsByOffset = (GcSlotId *)alloca(stackSlotCount * sizeof(GcSlotId));
+    // 0 is a valid slot id so default-initialize all the slots to 0xFFFFFFFF
+    memset(slotsByOffset, 0xFF, stackSlotCount * sizeof(GcSlotId));
+
+    INTERP_DUMP("Allocating gcinfo slots for %u vars\n", m_varsSize);
+
+    // Request slot IDs for all our vars, not just IL vars
+    for (int i = 0; i < m_varsSize; i++)
+    {
+        InterpVar *pVar = &m_pVars[i];
+        if (pVar->interpType != InterpTypeO)
+            continue;
+
+        // Globals are untracked and shouldn't need liveness ranges recorded in the info
+        GcSlotFlags flags = (GcSlotFlags)(pVar->global ? GC_SLOT_UNTRACKED : 0);
+        uint32_t slotIndex = pVar->offset / INTERP_STACK_SLOT_SIZE;
+        if (slotsByOffset[slotIndex] == ((GcSlotId)-1))
+        {
+            // Important to pass GC_xxx_REL, the default is broken due to GET_CALLER_SP being unimplemented
+            slotsByOffset[slotIndex] = gcInfoEncoder->GetStackSlotId(pVar->offset, flags, GC_FRAMEREG_REL);
+            INTERP_DUMP(
+                "Allocated gcinfo slot %u for %svar #%d at offset %d\n",
+                slotsByOffset[slotIndex], pVar->global ? "global " : "",
+                i, pVar->offset
+            );
+        }
+        else
+        {
+            INTERP_DUMP(
+                "Reused gcinfo slot %u for %svar #%d at offset %d\n",
+                slotsByOffset[slotIndex], pVar->global ? "global " : "",
+                i, pVar->offset
+            );
+            assert(!pVar->global);
+        }
+    }
+
+    gcInfoEncoder->FinalizeSlotIds();
+
+    // Now that slot IDs are finalized we want to record live ranges for every var that isn't global
+    for (int i = 0; i < m_varsSize; i++)
+    {
+        InterpVar *pVar = &m_pVars[i];
+        // Even if we have a gc slot for this offset, this var might not be an object reference
+        if (pVar->interpType != InterpTypeO)
+            continue;
+        // We don't need to report liveness ranges for untracked vars, the gc info decoder will report them
+        //  unconditionally.
+        if (pVar->global)
```

Comment:
If we don't report exact ranges for global vars, I think this would mean that we have to always zero them at method entry (even if they are not IL locals, even if they are not marked with initlocal). Otherwise GC could see some random uninitialized value.

Reply:
OK, it sounds like we need to report liveness ranges for everything then. That's a pretty straightforward change.

Reply:
Not sure how straightforward it would be. Let's say we have a global var used in the following way:
```
BB0:
     br BB2;
BB1:
    print global_var;
    ret;
BB2:
    br BB3;
BB3:
    global_var = 1;
    br BB1;
```

global_var is alive in BB1 and BB3 only. I don't think we currently have the machinery to track it in this fashion. Our current live ranges only make sense in the context of a single basic block.

Reply:
Is it difficult to ensure reftype globals have their storage pre-zeroed? I'm not sure what other options we have for the scenario you described.

Reply:
In our case we could trivially fix this by expanding the `INTOP_INITLOCALS` to zero also these vars (which follow immediately after the IL locals anyway). I would assume, in such cases, ryujit would report multiple live ranges for a single var ? @jakobbotsch ?

Reply:
Correct, RyuJIT will track the liveness of stack locals on per basic-block fashion. So in the above example, `global_var` would be live into BB1, but not into BB0, BB2 or BB3.

Reply:
Do you think I need to modify INITLOCALS in this PR? Or can we do it separately?

Reply:
I think we need to do it as part of this PR, to prevent any crashes, but it is trivial to do. You can save the end offset of the global vars space after by checking the value of `m_totalVarsStackSize` after `InitializeGlobalVars` is run. Then you can move the INITLOCAL to be inserted at the start of the entry bblock after the var offset allocator ran, or just have the end offset be updated when we emit the final IR as we do with offsets for vars.

Reply:
I modified things so that we always insert an initlocals opcode (with a size of 0 if initlocals wasn't needed), and record the opcode so I can expand it after offset allocation runs.